handlers/handler-router/src/main/java/org/apache/servicecomb/router/distribute/AbstractRouterDistributor.java

```diff
@@ -0,0 +1,176 @@
+      LOGGER.debug("route management can not match any rule and route the latest version");
+      return getLatestVersionList(list, targetServiceName);
+    }
+
+    TagItem targetTag = getFiltedServerTagItem(invokeRule, targetServiceName);
+    if (versionServerMap.containsKey(targetTag)) {
+      return versionServerMap.get(targetTag);
+    }
+    return getLatestVersionList(list, targetServiceName);
+  }
+
+  @Override
+  public void init(Function<T, E> getIns,
+      Function<E, String> getVersion,
+      Function<E, String> getServerName,
+      Function<E, Map<String, String>> getProperties) {
+    this.getIns = getIns;
+    this.getVersion = getVersion;
+    this.getServerName = getServerName;
+    this.getProperties = getProperties;
+  }
+
+  public TagItem getFiltedServerTagItem(PolicyRuleItem rule, String targetServiceName) {
+    return RouterRuleCache.getServiceInfoCacheMap().get(targetServiceName)
+        .getNextInvokeVersion(rule);
+  }
+
+  /**
+   * 1.filter targetService
+   * 2.return map distributed by version and tags
+   * establish map is a more complicate way than direct traversal，
+   * because we need to consider multiple matches，
+   * getProperties contains other irrelevant field except tags.
+   *
+   * @param serviceName
+   * @param list
+   * @return
+   */
+  private Map<TagItem, List<T>> getDistributList(String serviceName,
+      List<T> list,
+      PolicyRuleItem invokeRule) {
+    String latestV = RouterRuleCache.getServiceInfoCacheMap().get(serviceName).getLatestVersionTag()
+        .getVersion();
+    Map<TagItem, List<T>> versionServerMap = new HashMap<>();
+    for (T server : list) {
+      //get server
+      E ms = getIns.apply(server);
+      if (getServerName.apply(ms).equals(serviceName)) {
+        //most matching
+        TagItem tagItem = new TagItem(getVersion.apply(ms), getProperties.apply(ms));
+        TagItem targetTag = null;
+        int maxMatch = 0;
+        for (RouteItem entry : invokeRule.getRoute()) {
+          int nowMatch = entry.getTagitem().matchNum(tagItem);
+          if (nowMatch > maxMatch) {
+            maxMatch = nowMatch;
+            targetTag = entry.getTagitem();
+          }
+        }
+        synchronized (invokeRule) {
+          if (invokeRule.isWeightLess() && getVersion.apply(ms).equals(latestV)) {
+            TagItem latestVTag = invokeRule.getRoute().get(invokeRule.getRoute().size() - 1)
+                .getTagitem();
+            if (!versionServerMap.containsKey(latestVTag)) {
+              versionServerMap.put(latestVTag, new ArrayList<>());
+            }
+            versionServerMap.get(latestVTag).add(server);
+          }
+          if (targetTag != null) {
+            if (!versionServerMap.containsKey(targetTag)) {
+              versionServerMap.put(targetTag, new ArrayList<>());
+            }
+            versionServerMap.get(targetTag).add(server);
+          }
+        }
+      }
+    }
```

Comment:
I can remove the synchronized lock and make  `org.apache.servicecomb.router.model.ServiceInfoCache#allrule` to CopyOnWriteArrayList , but i can't decide which one is better. Do you guys have any suggestions?